src/codegen/traverse.lisp

```diff
@@ -250,9 +257,55 @@ other nodes, then it would be inappropriate to also define an action
               (fire-action ':traverse (class-name (class-of current-node)) actions args)
               (fire-action ':after    (class-name (class-of current-node)) actions args)
               (fire-action ':after    'node                                actions args))))
-      (let ((*traverse* #'current-traverse))
+      (let ((*traverse* (lambda (node &rest args)
+                          (funcall *traverse-hook*
+                                   (lambda ()
+                                     (apply #'current-traverse node args))
+                                   node))))
         (apply *traverse* initial-node initial-args)))))
 
+(defun identity-traverse-hook (thunk node)
+  "Default traversal hook function, which doesn't do any extra operation.
+This is called when the caller does _not_ use `traverse-with-hook`."
+  (declare (ignore node)) (funcall thunk))
+
+(defvar *traverse-hook* #'identity-traverse-hook
+  "A private dynamic variable used by `traverse-with-hook`.  This is rebound
+dynamically during running `traverse-with-hook`.")
+
+(defun traverse-with-hook (hook node action-list &rest args)
+  "Like `traverse`, but uses `hook` as the traversal hook function.
+The hook function is called whenever `traverse` recurse into the sub
+nodes.  It takes two arguments: `thunk` to recurse traversal, and `node`
+that is about to be recursed.
+
+The `traverse` function is equivalent to this with passing
+`identity-traverse-hook` as the hook.
```

Comment:
Right.  I didn't think the case that `traverse` is called during other `traverse`-family functions, but it is certainly possible.
Is there already a case that a function called from an action recursively calls `traverse`, thus can be affected with this commit? 
Otherwise, I'd just fix this docstring.

Reply:
Yes, the inliner definitely needs other sub-traversals of the code to be inlined, in order to rename type variables to fresh ones, as well as to perform AST variable substitution.